### PR TITLE
SAML Bug Fixes: Metadata Path, Resource Checking, AuthnRequest signature etc

### PR DESCRIPTION
**Pull request by mmoayyed**
This is a WIP, but I'll continue to push so you could see what changes are being added.

This closes hopefully #646 #645 #644 #643

### CODE REVIEW

**Comment by mmoayyed** [comment]
I am testing this with CAS5 RC2 running as a SAML IdP, and I realized redirect binding isnt quite supported with signed authn requests. I have added an option in pac4j to allow then, and will continue with CAS to handle signature values as request parameters.

**Comment by leleuj** [comment]
Thanks for taking these ones. Just let me know when you're done for the code review.

**Comment by mmoayyed** [comment]
I was able to figure out why pac4j isn't signing the requests. That bit is done and is now configurable. I also need to fix the CAS module to let it recognize the signature of the request for redirect bindings. However, once I do, I need your help to figure out how to run tests that produce a POST rather than a GET. I suppose you can even do that independant of this PR here and when it's ready, I'll merge and test it here.

**Comment by leleuj** [comment]
Are you referring to a POST (vs GET) towards the SAML IdP? In that case, indeed, it's a problem because the manual test is made for the user to copy/paste an url in the navigator, but we can output the HTML instead of the redirection url by changing things here: https://github.com/pac4j/pac4j/blob/master/pac4j-core/src/test/java/org/pac4j/core/run/RunClient.java#L37

Do your best, I'll finish things if necessary.

BTW, I hope to finish the new documentation before you finish this PR, so that you can review, complement and improve the SAML documentation.

**Comment by mmoayyed** [comment]
I have opted to bail on the SAML IT tests for now. I may do that bit in a separate pull request, given this PR is already complicated as it. Here is a quick update:

1. Bumped OpenSAML version
2. Made sure metadata resource existence is properly handled and checked.
3. Made sure authn requests are properly XML signed if needed.
4. Made sure metadata paths are correctly created if needed.

I'll wait for Travis to confirm.

**Comment by leleuj** [comment]
Travis is unhappy: "You have 3 PMD violations."

Apart from that, I finally added the documentation into the pac4j project: can you update this documentation with your changes: https://github.com/pac4j/pac4j/blob/master/documentation/docs/clients/saml.md ?

Reviewing your changes...

File: pac4j-core/src/main/java/org/pac4j/core/io/Resource.java [hunk 1]
```diff
@@ -1,21 +1,21 @@
...
 
+import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
 
 /**
  * Interface for a resource descriptor that abstracts from the actual
  * type of underlying resource, such as a file or class path resource.
- *
+ * <p>
```
Comment (leleuj) [inline pac4j-core/src/main/java/org/pac4j/core/io/Resource.java]: I'm surprised that the Javadoc task supports unclosed `<p>` tags, but let the build says if it likes it or not.

File: pac4j-core/src/main/java/org/pac4j/core/io/SpringResourceWrapper.java [hunk 2]
```diff
@@ -31,4 +37,13 @@ public String getFilename() {
 		return springResource.getFilename();
 	}
 
+	@Override
+	public File getFile() {
+		try {
+			return springResource.getFile();
+		} catch (final Exception e) {
+			LOGGER.error(e.getMessage(), e);
```
Comment (leleuj) [inline pac4j-core/src/main/java/org/pac4j/core/io/SpringResourceWrapper.java]: It's a not a big deal, but shouldn't we throw a `TechnicalException` instead?

File: pac4j-saml/src/main/java/org/pac4j/saml/client/SAML2ClientConfiguration.java [hunk 3]
```diff
@@ -40,6 +40,8 @@
 
     private boolean forceAuth = false;
 
+    private boolean forceSignRedirectBindingAuthnRequest;
```
Comment (leleuj) [inline pac4j-saml/src/main/java/org/pac4j/saml/client/SAML2ClientConfiguration.java]: These new options will need appropriate documentation ;-)

File: pac4j-saml/src/main/java/org/pac4j/saml/client/SAML2ClientConfiguration.java [hunk 4]
```diff
@@ -63,7 +65,8 @@ public SAML2ClientConfiguration() {}
     private List<String> signatureReferenceDigestMethods;
     private String signatureCanonicalizationAlgorithm;
     private boolean wantsAssertionsSigned = true;
-
+    private boolean signErrorResponses;
```
Comment (leleuj) [inline pac4j-saml/src/main/java/org/pac4j/saml/client/SAML2ClientConfiguration.java]: What's the meaning of this new property?

File: pac4j-saml/src/main/java/org/pac4j/saml/metadata/SAML2ServiceProviderMetadataResolver.java [hunk 5]
```diff
@@ -126,7 +127,11 @@ public final MetadataResolver resolve() {
                     logger.info("Metadata file already exists at {}.", this.spMetadataResource.getFilename());
                 } else {
                     logger.info("Writing sp metadata to {}", this.spMetadataResource.getFilename());
-
+                    final File parent = spMetadataResource.getFile().getParentFile();
+                    if (!parent.mkdirs()) {
+                        logger.error("Could not construct the directory structure for SP metadata {}",
```
Comment (leleuj) [inline pac4j-saml/src/main/java/org/pac4j/saml/metadata/SAML2ServiceProviderMetadataResolver.java]: Shouldn't we throw a `TechnicalException` here as we were unable to create the directory where the SP metadata will be written?

File: pac4j-saml/src/main/java/org/pac4j/saml/sso/impl/SAML2WebSSOMessageSender.java [hunk 6]
```diff
@@ -119,7 +123,7 @@ protected final void invokeOutboundMessageHandlers(final SPSSODescriptor spDescr
                 final SAMLOutboundProtocolMessageSigningHandler handler = new
                         SAMLOutboundProtocolMessageSigningHandler();
                 handler.setSignErrorResponses(this.signErrorResponses);
-                handler.invoke(outboundContext);
+                //handler.invoke(outboundContext);
```
Comment (leleuj) [inline pac4j-saml/src/main/java/org/pac4j/saml/sso/impl/SAML2WebSSOMessageSender.java]: I'm surprised to see that line commented. I had in mind that this was were the work was done...

File: pac4j-saml/src/test/java/org/pac4j/saml/sso/impl/SAML2DefaultResponseValidatorTests.java [hunk 7]
```diff
@@ -1,100 +0,0 @@
...
-import org.opensaml.saml.saml2.metadata.SPSSODescriptor;
-import org.pac4j.saml.context.SAML2MessageContext;
-import org.pac4j.saml.crypto.SAML2SignatureTrustEngineProvider;
-import org.pac4j.saml.exceptions.SAMLException;
-
-import static org.junit.Assert.*;
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.when;
-
-public class SAML2DefaultResponseValidatorTests {
```
Comment (leleuj) [inline pac4j-saml/src/test/java/org/pac4j/saml/sso/impl/SAML2DefaultResponseValidatorTests.java]: This test class is no longer necessary?

**Comment by leleuj** [comment]
A few comments here and there, but certainly mostly because of my low skills on SAML...

+1

File: pac4j-core/src/main/java/org/pac4j/core/io/Resource.java (hunk 1 above)
Comment (mmoayyed) [inline pac4j-core/src/main/java/org/pac4j/core/io/Resource.java]: Looks like Travis is skipping Javadocs: `mvn install -DskipTests=true -Dmaven.javadoc.skip=true -B -V`

Comment (leleuj) [inline pac4j-core/src/main/java/org/pac4j/core/io/Resource.java]: Yes, but I'm pretty sure I won't when I'll perform the release ;-)

**Comment by leleuj** [comment]
Unfortunately, tests are broken due to the fact we can't create write the SP metadata...

**Comment by mmoayyed** [comment]
I am playing around with Travis to see why tests pass locally for me but not Travis! This will eventually pass at some point.

**Comment by leleuj** [comment]
Excellent! I'm ready to merge. Anything to add?

I'll reopen #643 and do it myself.

**Comment by mmoayyed** [comment]
Sounds good. Go ahead. I'll take a pass at the docs tomorrow as well (with whatever updates this PR has) and see how things look.

**Pull request merged**